BaikalDataGenerator/Source/render.cpp

```diff
@@ -0,0 +1,351 @@
+    m_renderer = m_factory->CreateRenderer(Baikal::ClwRenderFactory::RendererType::kUnidirectionalPathTracer);
+    m_controller = m_factory->CreateSceneController();
+
+    for (auto& output_info: outputs_collection)
+    {
+        m_outputs.push_back(m_factory->CreateOutput(output_width, output_height));
+        m_renderer->SetOutput(output_info.type, m_outputs.back().get());
+        output_info.width = output_width;
+        output_info.height = output_height;
+    }
+
+    if (!std::filesystem::exists(scene_file))
+    {
+        THROW_EX("There is no any scene file to load");
+    }
+
+    m_scene = Baikal::SceneIo::LoadScene(scene_file.string(),
+                                         scene_file.parent_path().string());
+}
+
+void Render::UpdateCameraSettings(const CameraInfo& cam_state)
+{
+    if (cam_state.aperture != m_camera->GetAperture())
+    {
+        m_camera->SetAperture(cam_state.aperture);
+    }
+
+    if (cam_state.focal_length != m_camera->GetFocalLength())
+    {
+        m_camera->SetFocalLength(cam_state.focal_length);
+    }
+
+    if (cam_state.focus_distance != m_camera->GetFocusDistance())
+    {
+        m_camera->SetFocusDistance(cam_state.focus_distance);
+    }
+
+    auto cur_pos = m_camera->GetPosition();
+    auto at = m_camera->GetForwardVector();
+    auto up = m_camera->GetUpVector();
+
+    if (cur_pos != cam_state.pos ||
+        at != cam_state.at ||
+        up != cam_state.up)
+    {
+        m_camera->LookAt(cam_state.pos, cam_state.at, cam_state.up);
+    }
+}
+
+void Render::SaveOutput(OutputInfo desc,
+                        int cam_index,
+                        int spp,
+                        bool gamma_correction_enabled,
+                        const std::filesystem::path& output_dir)
+{
+    OIIO_NAMESPACE_USING;
+
+    auto output = m_renderer->GetOutput(desc.type);
+
+    assert(output);
+
+    auto buffer = static_cast<Baikal::ClwOutput*>(output)->data();
+
+    std::vector<RadeonRays::float3> output_data(buffer.GetElementCount());
+    std::vector<RadeonRays::float3> image_data(buffer.GetElementCount());
+
+    output->GetData(output_data.data());
+
+    auto width = output->width();
+    auto height = output->height();
+
+    if (gamma_correction_enabled)
+    {
+        for (auto y = 0u; y < height; ++y)
+        {
+            for (auto x = 0u; x < width; ++x)
+            {
+                float3 val = output_data[(height - 1 - y) * width + x];
+                // "The 4-th pixel component is a count of accumulated samples.
+                //It can be different for every pixel in case of adaptive sampling.
+                //So, we need to normalize pixel values here".
+                val *= (1.f / val.w);
+                // gamma corection
+                image_data[y * width + x].x = std::pow(val.x, 1.f / 2.2f);
+                image_data[y * width + x].y = std::pow(val.y, 1.f / 2.2f);
+                image_data[y * width + x].z = std::pow(val.z, 1.f / 2.2f);
```

Comment:
Also Depth and Gloss have only on channel as far as I know. Why do we calculate 3 values ?